src/pages/profile/edit.tsx

```diff
@@ -0,0 +1,660 @@
+  const [handle, setHandle] = useState(initUser.handle);
+  const [email, setEmail] = useState(initUser.email);
+  const [major, setMajor] = useState(initUser.major);
+  const [graduationYear, setGraduationYear] = useState(String(initUser.graduationYear));
+  const [bio, setBio] = useState(initUser.bio);
+
+  const [isAttendancePublic, setIsAttendancePublic] = useState(initUser.isAttendancePublic);
+  const [isResumeVisible, setIsResumeVisible] = useState(
+    initUser.resumes?.some(resume => resume.isResumeVisible) || initUser.resumes?.length === 0
+  );
+
+  const [password, setPassword] = useState('');
+  const [newPassword, setNewPassword] = useState('');
+  const [confirmPassword, setConfirmPassword] = useState('');
+
+  const [socialMedia, setSocialMedia] = useState(
+    () => new Map(initUser.userSocialMedia?.map(social => [social.type, social.url]) ?? [])
+  );
+
+  const hasChange =
+    firstName !== user.firstName ||
+    lastName !== user.lastName ||
+    handle !== user.handle ||
+    email !== user.email ||
+    major !== user.major ||
+    +graduationYear !== user.graduationYear ||
+    bio !== user.bio ||
+    isAttendancePublic !== user.isAttendancePublic ||
+    Array.from(socialMedia).some(
+      ([type, url]) =>
+        (user.userSocialMedia?.find(social => social.type === type)?.url ?? '') !== url
+    ) ||
+    newPassword.length > 0;
+
+  // Warn if there are unsaved changes
+  useEffect(() => {
+    if (hasChange) {
+      const handleBeforeUnload = (e: BeforeUnloadEvent) => {
+        e.preventDefault();
+        e.returnValue = true;
+      };
+      window.addEventListener('beforeunload', handleBeforeUnload);
+      return () => {
+        window.removeEventListener('beforeunload', handleBeforeUnload);
+      };
+    }
+    return undefined;
+  }, [hasChange]);
+
+  const [resumes, setResumes] = useState(initUser.resumes ?? []);
+
+  const years = useMemo(() => {
+    const years: number[] = [];
+    const currentYear = new Date().getUTCFullYear();
+    // Ensure the user's graduation year is listed in the dropdown in case
+    // they're taking longer than expected to graduate
+    for (
+      let year = Math.min(user.graduationYear, currentYear - 2);
+      year <= currentYear + 6;
+      year += 1
+    ) {
+      years.push(year);
+    }
+    return years;
+  }, [user.graduationYear]);
+
+  return (
+    <>
+      <h1 className={styles.title}>Edit Profile</h1>
+      <div className={styles.mainContent}>
+        <div className={styles.columns}>
+          <section className={styles.columnLeft}>
+            <h2>Current Profile</h2>
+            <Preview user={user} pfpCacheBust={pfpCacheBust} />
+          </section>
+          <form
+            className={styles.columnRight}
+            onSubmit={async e => {
+              e.preventDefault();
+
+              let changed = false;
+              if (
+                firstName !== user.firstName ||
+                lastName !== user.lastName ||
```

Comment:
Thoughts on putting this into a `handleSubmit` function to make the rest of the JSX easier to read?